File: src/com/suushiemaniac/lang/json/value/JsonString.java
package com.suushiemaniac.lang.json.value;

import com.suushiemaniac.lang.json.JSON;
import com.suushiemaniac.lang.json.util.StringUtils;

public class JsonString extends JsonType {
    private String value;

    public JsonString(String value) {
        this(null, value);
    }

    public JsonString(JSON parent, String value) {
        super(parent);
        this.value = StringUtils.jsonUnwrap(value);
    }

    @Override
    public String stringValue() {
        return this.value;
    }

    @Override
    public Object toNative() {
        return this.stringValue();
    }

    @Override
    public boolean booleanValue() {
        return this.value.length() > 0;
    }

    @Override
    public String toString() {
        return StringUtils.jsonWrap(this.value);
    }

    @Override
    public boolean equals(Object other) {
        return (other instanceof JsonString && ((JsonString) other).value.equals(this.value))
                || (other instanceof String && other.equals(this.value));
    }

    @Override
    public void set(JSON value) {
        if (value instanceof JsonString) this.value = value.stringValue();
    }
}

File: src/main/kotlin/com/suushiemaniac/lang/json/value/JsonString.kt
package com.suushiemaniac.lang.json.value

import com.suushiemaniac.lang.json.JSON
import com.suushiemaniac.lang.json.util.StringUtils.jsonUnwrap
import com.suushiemaniac.lang.json.util.StringUtils.jsonWrap

class JsonString(parent: JSON?, var value: String) : JsonType(parent) {
    constructor(value: String) : this(null, value)

    init {
        this.value = value.jsonUnwrap()
    }

    override fun stringValue(): String {
        return this.value
    }

    override fun toNative(): Any {
        return this.stringValue()
    }

    override fun booleanValue(): Boolean {
        return this.value.isNotEmpty()
    }

    override fun intValue(): Int {
        return try {
            this.stringValue().toInt()
        } catch (e: NumberFormatException) {
            super.intValue()
        }

    }

    override fun floatValue(): Float {
        return try {
            this.value.toFloat()
        } catch (e: NumberFormatException) {
            super.floatValue()
        }
    }

    override fun toString(): String {
        return this.value.jsonWrap()
    }

    override fun equals(other: Any?): Boolean {
        return other is JsonString && other.value == this.value || other is String && other == this.value
    }

    override fun hashCode(): Int {
        return this.value.hashCode()
    }

    override fun set(value: JSON) {
        if (value is JsonString) this.value = value.stringValue()
    }
}
